### PR TITLE
Add parsing of double values in style attributes

### PR DESCRIPTION
**Pull request by KiChjang**
This change is [Reviewable](https://reviewable.io/reviews/servo/servo/10608)

### CODE REVIEW

**Comment by KiChjang** [comment]
```
$ bash etc/ci/check_no_unwrap.sh

components/compositing/constellation.rs:870:                          LoadData::new(Url::parse("about:failure").unwrap()));

components/compositing/constellation.rs:953:                .unwrap_or_else(|| Url::parse("about:blank").unwrap());

The command "bash etc/ci/check_no_unwrap.sh" exited with 1.
```
Not my fault!

**Comment by jdm** [comment]
Yep, #10610 clears that up.

File: components/style/attr.rs [hunk 1]
```diff
@@ -73,6 +74,50 @@ pub fn parse_unsigned_integer<T: Iterator<Item=char>>(input: T) -> Option<u32> {
     })
 }
 
+/// Parse a floating-point number according to
+/// <https://html.spec.whatwg.org/multipage/#rules-for-parsing-floating-point-number-values>
+pub fn parse_double(string: &DOMString) -> Option<f64> {
```
Comment (nox) [inline components/style/attr.rs]: Nit: could it return `Result<f64, ()>` instead?

**Comment by nox** [comment]
Shouldn't that make new tests pass?

-S-awaiting-review +S-needs-code-changes 

---

Reviewed 7 of 7 files at r1.
Review status: all files reviewed at latest revision, 7 unresolved discussions.

---

*[components/servo/Cargo.lock, line 2193 \[r1\]](https://reviewable.io:443/reviews/servo/servo/10608#-KFOdic6OGEdjoyciptF:-KFOdic6OGEdjoyciptG:1094760345) ([raw file](https://github.com/servo/servo/blob/d86b5eb902dff4374f6034fc608336d7101ac895/components/servo/Cargo.lock#L2193)):*
You should probably use `./mach cargo-update`, I'm pretty sure more lock files need to change.

---

*[components/style/attr.rs, line 82 \[r1\]](https://reviewable.io:443/reviews/servo/servo/10608#-KFOg_-8xzFRy2e0vMKM:-KFOg_-8xzFRy2e0vMKN:-1092715673) ([raw file](https://github.com/servo/servo/blob/d86b5eb902dff4374f6034fc608336d7101ac895/components/style/attr.rs#L82)):*
Just use `trim_matches` to skip whitespace.

---

*[components/style/attr.rs, line 106 \[r1\]](https://reviewable.io:443/reviews/servo/servo/10608#-KFOgFVpOO0HfB8zm47f:-KFOgFVpOO0HfB8zm47g:-2015538923) ([raw file](https://github.com/servo/servo/blob/d86b5eb902dff4374f6034fc608336d7101ac895/components/style/attr.rs#L106)):*
No need to recreate the iterator.

---

*[components/util/str.rs, line 166 \[r1\]](https://reviewable.io:443/reviews/servo/servo/10608#-KFOiwjcNCp97ZLbOaKO:-KFOiwjdlnQ8TQPZdWPJ:-1785858353) ([raw file](https://github.com/servo/servo/blob/d86b5eb902dff4374f6034fc608336d7101ac895/components/util/str.rs#L166)):*
Nit: can you make this a char?

---

*[components/util/str.rs, line 170 \[r1\]](https://reviewable.io:443/reviews/servo/servo/10608#-KFOiyNeTT9vlSm73jFo:-KFOiyNeTT9vlSm73jFp:-1702940532) ([raw file](https://github.com/servo/servo/blob/d86b5eb902dff4374f6034fc608336d7101ac895/components/util/str.rs#L170)):*
Why not just do `c == '.'` btw?

---

*[components/util/str.rs, line 196 \[r1\]](https://reviewable.io:443/reviews/servo/servo/10608#-KFOhY_dLecx0ZvxtPyo:-KFOhY_dLecx0ZvxtPyp:-587549141) ([raw file](https://github.com/servo/servo/blob/d86b5eb902dff4374f6034fc608336d7101ac895/components/util/str.rs#L196)):*
Why do we need to return the number of digits which were read?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/servo/servo/10608#-:-KFOeLTbjDGBE4TM6UM5:1599058525)*

**Comment by KiChjang** [comment]
Review status: all files reviewed at latest revision, 7 unresolved discussions.

---

*[components/style/attr.rs, line 79 \[r1\]](https://reviewable.io:443/reviews/servo/servo/10608#-KFKJsMv9MhJ-XBcOqkR-r1-79:-KFPFU7z_5qpRxkepMas:-1410775169) ([raw file](https://github.com/servo/servo/blob/d86b5eb902dff4374f6034fc608336d7101ac895/components/style/attr.rs#L79)):*
Should I change parse_integer as well then?

---

*[components/style/attr.rs, line 106 \[r1\]](https://reviewable.io:443/reviews/servo/servo/10608#-KFOgFVpOO0HfB8zm47f:-KFPApWuXJIzXHVfiS4Z:-1846411197) ([raw file](https://github.com/servo/servo/blob/d86b5eb902dff4374f6034fc608336d7101ac895/components/style/attr.rs#L106)):*
Can't do that, the `read_*` functions take ownership of the iterator.

---

*[components/util/str.rs, line 196 \[r1\]](https://reviewable.io:443/reviews/servo/servo/10608#-KFOhY_dLecx0ZvxtPyo:-KFPIaBBMMjF5sRKMBok:-1167466664) ([raw file](https://github.com/servo/servo/blob/d86b5eb902dff4374f6034fc608336d7101ac895/components/util/str.rs#L196)):*
Because these functions take ownership of the iterator, and so I need to keep track of where it stopped parsing.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/servo/servo/10608)*

**Comment by nox** [comment]
-S-awaiting-review +S-needs-code-changes 

---

Reviewed 5 of 5 files at r2.
Review status: all files reviewed at latest revision, 5 unresolved discussions, some commit checks failed.

---

*[components/style/attr.rs, line 79 \[r1\]](https://reviewable.io:443/reviews/servo/servo/10608#-KFKJsMv9MhJ-XBcOqkR-r1-79:-KFcTD0SWMMOFSd5n6U3:2120061629) ([raw file](https://github.com/servo/servo/blob/d86b5eb902dff4374f6034fc608336d7101ac895/components/style/attr.rs#L79)):*
Would be nice yes.

---

*[components/style/attr.rs, line 106 \[r1\]](https://reviewable.io:443/reviews/servo/servo/10608#-KFOgFVpOO0HfB8zm47f:-KFcSwcv_lzxsTIyxkVm:-1289762518) ([raw file](https://github.com/servo/servo/blob/d86b5eb902dff4374f6034fc608336d7101ac895/components/style/attr.rs#L106)):*
Well, just pass `&mut Peekable<I>`, no?

---

*[components/util/str.rs, line 196 \[r1\]](https://reviewable.io:443/reviews/servo/servo/10608#-KFOhY_dLecx0ZvxtPyo:-KFcT5IIIaeOj-ea6IMw:213059064) ([raw file](https://github.com/servo/servo/blob/d86b5eb902dff4374f6034fc608336d7101ac895/components/util/str.rs#L196)):*
Just make them take `&mut Peekable<I>`.

---

*[components/util/str.rs, line 170 \[r2\]](https://reviewable.io:443/reviews/servo/servo/10608#-KFcT2ou19IOOx6Cyyf0:-KFcT2ou19IOOx6Cyyf1:-1675573302) ([raw file](https://github.com/servo/servo/blob/c0af9fe51c7b2d4357ad1555e4e133c8f97f6e56/components/util/str.rs#L170)):*
Nit: `char`.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/servo/servo/10608#-:-KFcTHmAS_GTPG7O1y6r:1458266033)*

**Comment by KiChjang** [comment]
Review status: all files reviewed at latest revision, 5 unresolved discussions, some commit checks failed.

---

*[components/style/attr.rs, line 106 \[r1\]](https://reviewable.io:443/reviews/servo/servo/10608#-KFOgFVpOO0HfB8zm47f:-KFdylawIjHbfmxI466m:2024456849) ([raw file](https://github.com/servo/servo/blob/d86b5eb902dff4374f6034fc608336d7101ac895/components/style/attr.rs#L106)):*
Ack, this actually causes my parsing to fail, because take_while() consumes the first element that doesn't satisfy the condition. I could use `take_while_ref` offered by the itertools crate though, should I use that?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/servo/servo/10608)*

**Comment by KiChjang** [comment]
Review status: all files reviewed at latest revision, 5 unresolved discussions, some commit checks failed.

---

*[components/style/attr.rs, line 106 \[r1\]](https://reviewable.io:443/reviews/servo/servo/10608#-KFOgFVpOO0HfB8zm47f:-KFe0zYFIzsdt_E3zGza:1243719068) ([raw file](https://github.com/servo/servo/blob/d86b5eb902dff4374f6034fc608336d7101ac895/components/style/attr.rs#L106)):*
Doesn't seem to work either, the iterator needs to be cloneable in order to use `.take_while_ref()`.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/servo/servo/10608)*

**Comment by nox** [comment]
Reviewed 3 of 3 files at r3.
Review status: all files reviewed at latest revision, 2 unresolved discussions, some commit checks failed.

---

*[components/style/attr.rs, line 106 \[r1\]](https://reviewable.io:443/reviews/servo/servo/10608#-KFOgFVpOO0HfB8zm47f:-KFsHSKFsdCdUGXxDCkw:-426213581) ([raw file](https://github.com/servo/servo/blob/d86b5eb902dff4374f6034fc608336d7101ac895/components/style/attr.rs#L106)):*
I don't see a `take_while` call anywhere so I'm confused.

Anyway I don't think having to decode a string multiple times should be needed to just parse an integer string.

---

*[components/style/attr.rs, line 66 \[r3\]](https://reviewable.io:443/reviews/servo/servo/10608#-KFsH6bLeXW2GO9zxtsx:-KFsH6bLeXW2GO9zxtsy:2061751729) ([raw file](https://github.com/servo/servo/blob/6d709cbe324c83031e5ed74b326215afd5065357/components/style/attr.rs#L66)):*
Shouldn't `do_parse_integer` return a `Result` itself? Feel free to file an E-Easy about that.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/servo/servo/10608)*

**Comment by KiChjang** [comment]
Review status: all files reviewed at latest revision, 2 unresolved discussions, some commit checks failed.

---

*[components/style/attr.rs, line 106 \[r1\]](https://reviewable.io:443/reviews/servo/servo/10608#-KFOgFVpOO0HfB8zm47f:-KFtpD1Ar6eziy-8G7nW:456459893) ([raw file](https://github.com/servo/servo/blob/d86b5eb902dff4374f6034fc608336d7101ac895/components/style/attr.rs#L106)):*
`take_while` is what's being used in `read_integers` and `read_fractions`. I don't think it's necessary either but due to the design of `take_while`,  I can't track the last character that wasn't a digit.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/servo/servo/10608)*

**Comment by nox** [comment]
Seems good enough. Please file an issue about `do_parse_integers`' return type and `take_while` being used.

No WPT test?

-S-awaiting-review +S-awaiting-answer 

---

Reviewed 10 of 10 files at r4.
Review status: all files reviewed at latest revision, 2 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/servo/servo/10608#-:-KGGsjXb7sqxg11bwsV4:1602744021)*

**Comment by nox** [comment]
Seems good enough. Please file an issue about `do_parse_integers`' return type and `take_while` being used.

No WPT test?

-S-awaiting-review +S-awaiting-answer 

---

Reviewed 10 of 10 files at r4.
Review status: all files reviewed at latest revision, 2 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/servo/servo/10608#-:-KGGsjXb7sqxg11bwsV4:1602744021)*

**Comment by KiChjang** [comment]
I have another branch that actually utilizes the parse_double utility function (it's for parsing HTMLInputElement doubles), but for this PR, I've added a unit test instead since it's only a util crate modification.

**Comment by KiChjang** [comment]
Filed #10860 and #10861.

@bors-servo r=nox

**Pull request merged**